1.Load the dataset and explore the variables.

In [1]:
import pandas as pd

In [2]:
dataset = pd.read_csv("files_for_lab/customer_churn.csv")

In [3]:
dataset.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [5]:
dataset.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

2.We will try to predict variable Churn using a logistic regression on variables tenure, SeniorCitizen,MonthlyCharges.
3.Split the Dataset into X ('tenure', 'SeniorCitizen', 'MonthlyCharges') and y ('Churn')

In [6]:
from sklearn.model_selection import train_test_split
X = dataset[['tenure', 'SeniorCitizen', 'MonthlyCharges']]
y = dataset['Churn']

3.Split the Dataset into X ('tenure', 'SeniorCitizen', 'MonthlyCharges') and y ('Churn')

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

4.Build the logistic regression model.

In [8]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)


LogisticRegression(random_state=42)

5.Evaluate the model.

In [9]:
from sklearn.metrics import accuracy_score, classification_report

In [10]:
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

In [11]:
print("Accuracy:", accuracy)
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.8055358410220014
Classification Report:
               precision    recall  f1-score   support

          No       0.83      0.92      0.87      1036
         Yes       0.69      0.48      0.57       373

    accuracy                           0.81      1409
   macro avg       0.76      0.70      0.72      1409
weighted avg       0.79      0.81      0.79      1409



6.Even a simple model will give us more than 70% accuracy. Why?


Simple model will give us more than 70% accuracy in few reasons, if the data is easy separable model may preform well if the data is well structured and distinct, also if we have meaningful features models will preform well. Data size is important if we have lagre datat set model can capture underlaying patterns and variations in data. 

7.Synthetic Minority Oversampling TEchnique (SMOTE) is an over sampling technique based on nearest neighbors that adds new points between existing points. Apply imblearn.over_sampling.SMOTE to the dataset. Build and evaluate the logistic regression model. Is it there any improvement?

In [12]:

conda install -c conda-forge imbalanced-learn

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\amraf\anaconda3

  added / updated specs:
    - imbalanced-learn


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    imbalanced-learn-0.11.0    |     pyhd8ed1ab_0         138 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         138 KB

The following packages will be UPDATED:

  imbalanced-learn   pkgs/main/win-64::imbalanced-learn-0.~ --> conda-forge/noarch::imbalanced-learn-0.11.0-pyhd8ed1ab_0 




imbalanced-learn-0.1 | 138 KB    |            |   0% 
imbalanced-learn-0.1 | 138 KB    | #1         |  12% 
imbalanced-learn-0.1 | 138 KB    | ########## | 100% 
imbalanced-learn-0.1 | 138 KB    | ########## | 100% 
                                                     


Preparing transaction: ...working... done
Verif

DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/r/win-64/current_repodata.json HTTP/1.1" 304 0
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/main/win-64/current_repodata.json HTTP/1.1" 304 0
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/msys2/win-64/current_repodata.json HTTP/1.1" 304 0
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/msys2/noarch/current_repodata.json HTTP/1.1" 304 0
DEBUG

In [13]:
from imblearn.over_sampling import SMOTE

In [14]:
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train, y_train)


In [15]:
model_smote = LogisticRegression(random_state=42)
model_smote.fit(X_resampled, y_resampled)


LogisticRegression(random_state=42)

In [16]:
y_pred_smote = model_smote.predict(X_test)
accuracy_smote = accuracy_score(y_test, y_pred_smote)


In [17]:
print("\nModel with SMOTE:")
print("Accuracy:", accuracy_smote)
print("Classification Report:\n", classification_report(y_test, y_pred_smote))


Model with SMOTE:
Accuracy: 0.7459190915542938
Classification Report:
               precision    recall  f1-score   support

          No       0.90      0.74      0.81      1036
         Yes       0.51      0.76      0.61       373

    accuracy                           0.75      1409
   macro avg       0.71      0.75      0.71      1409
weighted avg       0.80      0.75      0.76      1409



The accuracy of the model without SMOTE is higher than the model with SMOTE (0.8055 vs. 0.7459).Accuracy alone may not be the best metric, especially in imbalanced datasets.
The precision for predicting "Yes" (churn) is lower in the model with SMOTE (0.51) compared to the model without SMOTE (0.69).

The model with SMOTE has a higher recall for "Yes" (0.76) compared to the model without SMOTE (0.48).

F1-Score for "Yes" is higher in the model with SMOTE (0.61) compared to the model without SMOTE (0.57).

The model with SMOTE shows improvement in capturing positive instances (churn), as seen in the higher recall and F1-Score for the "Yes" class , but the accuracy has decreased. 